### PR TITLE
Android accessibility-label alignment with iOS

### LINKED ISSUE #3977: Text property/matching: Derive from child-views
### Describe your idea

As reported via an [internal support case](https://wix.slack.com/archives/C1Y4D9BK8/p1678275248117969) by @bunkib, we could do better in terms of Android equivalence (to iOS) by aligning text-based matching logic:

On iOS, the _text_ property of a view is defined not only by the text associated with view itself - provided that it is in fact a text view, but rather by recursively concatenating the text values of all children. This is not because this is how we decided to introduce that behavior on iOS, but because this is how React Native sees it (IIRC).

This is relevant on 2 levels:
1. `.toHaveText()` matching logic: Look up the text based on the view's textual children, recursively.
2. `text` attribute: Assign based on the children's texts, similarly.

### PR DESCRIPTION
**Pull request by d4vidi**
## Description

- This pull request fixes #3977

In this pull request, I have applied a more extensive algorithm for resolving a `View`'s `contentDescription` so as to bridge over the RN gap in https://github.com/facebook/react-native/issues/32826 (as explained in #3977).

---

> _For features/enhancements:_
 - [x] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [x] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/ui/UIExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        if (rawLabel != null) {
+            subLabels.add(rawLabel)
+        } else {
+            view.forEachChild { child ->
+                collectChildAccessibilityLabels(child, subLabels)
+            }
+        }
+    }
+    return subLabels
+}
```
Comment (asafkorem) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/ui/UIExtensions.kt:32]: You could avoid this recursion using a queue:
```kotlin
private fun collectChildAccessibilityLabels(
    view: View,
): List<CharSequence> {
    val labels = mutableListOf<CharSequence>()
    val queue = mutableListOf<View>()

    queue.add(view)

    while (queue.isNotEmpty()) {
        val currentView = queue.removeAt(0)
        getRawAccessibilityLabel(currentView)?.let { rawLabel ->
            labels.add(rawLabel)
        } ?: run {
            view.forEachChild { child ->
                queue.add(child)
            }
        }
    }

    return labels
}

```

File: detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxMatcher.java [hunk 2]
```diff
@@ -47,6 +48,14 @@ public static Matcher<View> matcherForContentDescription(String contentDescripti
         return allOf(withContentDescription(contentDescription), withEffectiveVisibility(Visibility.VISIBLE));
     }
 
+    public static Matcher<View> matcherForContentDescriptionOrText(String contentDescription) {
+        return allOf(anyOf(withContentDescription(contentDescription), withText(contentDescription)), withEffectiveVisibility(Visibility.VISIBLE));
+    }
```
Comment (asafkorem) [inline detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxMatcher.java:53]: I don't really get what you're doing here to be honest. Let's discuss this part..

**Review by asafkorem — CHANGES_REQUESTED** [review]
image

File: detox/src/android/matchers/index.js [hunk 3]
```diff
@@ -2,9 +2,9 @@ const native = require('./native');
 const web = require('./web');
 
 module.exports = {
-  accessibilityLabel: (value) => new native.LabelMatcher(value),
   id: (value) => new native.IdMatcher(value),
-  label: (value) => new native.LabelMatcher(value),
+  label: (value) => new native.ExtendedValueMatcher(value),
+  accessibilityLabel: (value) => new native.ExtendedValueMatcher(value),
```
[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/src/android/matchers/index.js:7]: I don't know either - `accessibilityLabel` isn't even documented. I kept it for the benefit of the doubt that it is used by someone, somewhere. We can deprecated for Detox 21/22

[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/src/android/matchers/index.js:7]: ... https://github.com/wix/Detox/discussions/3406#discussion-4081876

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/ui/UIExtensions.kt (hunk 1 above)
Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/ui/UIExtensions.kt:32]: I liked the idea of traversing the view hierarchy as a general method.

File: detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxMatcher.java (hunk 2 above)
Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxMatcher.java:53]: I'll try to improve conventions

**Review by asafkorem — APPROVED** [review]

**Pull request merged**